scripts/search/rank.js

```diff
@@ -0,0 +1,160 @@
+module.exports = {
+
+    // Return the explicit or implicit rank for a given page. See the Algolia documentation to learn more about how Algolia indexing works.
+    // https://www.algolia.com/doc/guides/managing-results/must-do/custom-ranking/
+    get(page) {
+
+        // If the page is explicitly ranked zero, return zero.
+        const explicitRank = this.getExplicitRank(page);
+        if (explicitRank === 0) {
+            return 0;
+        }
+
+        // If it's ranked otherwise numerically, return that.
+        if (!isNaN(explicitRank)) {
+            return explicitRank;
+        }
+
+        // Otherwise, return its implicit rank.
+        return this.getImplicitRank(page);
+    },
+
+    // Rank can be set at the page level as well (e.g., in YAML frontmatter: `search.rank: 123`).
+    getExplicitRank(page) {
+        return page.params?.search?.rank;
+    },
+
+    // Pages can be boosted at the page level also (e.g., `search.boost: true`)
+    isExplicitlyBoosted(page) {
+        return !!page.params?.search?.boost;
+    },
+
+    // Implicit keywords are derived from the page's URL. We do this to make string matches more
+    // likely. (For example, to surface both AWS Classic and AWS Native for queries for 'aws'.)
+    getImplicitKeywords(page) {
+        const match = page.href.match(/^\/registry\/packages\/(aws|aws-native|azure|azure-native|gcp|google-native|kubernetes)\/$/);
+
+        if (match && match[1]) {
+            return [
+                match[1].replace("-native", ""),
+            ];
+        }
+
+        return [];
+    },
+
+    // Implicit rankings allow us to codify how relevant our pages are in relation to one another.
+    // Rankings are applied in descending order numerically (i.e., for a given query, higher-ranked
+    // pages will will show up before lower-ranked ones).
+    getImplicitRank(page) {
+
+        // Get-started pages.
+        if (page.href.match(/^\/docs\/clouds\/(aws|azure|gcp|kubernetes)\/get-started\//)) {
+
+            // Top-level get-started pages rank highest.
+            if (page.kind === "section") {
+                return 1000;
+            }
+
+            // All other get-started pages get de-listed. We don't want them noising up the results,
+            // as they aren't the kinds of pages you'd typically want to navigate to directly.
+            return 0;
+        }
+
+        // Cloud landing pages are next.
+        if (page.href.match(/^\/docs\/clouds\/(aws|azure|gcp|kubernetes)\//)) {
+
+            // Top-level pages again rank highest.
+            if (page.kind === "section") {
+                return 990;
+            }
+
+            return 980;
+        }
+
+        // Docs pages are next.
```

Comment:
i didnt have a chance to bring this up in the meeting but we get way more daily visitors to registry than docs. should registry come before docs?

Reply:
I reviewed GA as part of this, yeah, and based on all the learning and experimentation I've done to this point, I'm pretty sure the answer to this question is no, because Algolia custom rankings aren't really meant to convey popularity; they exist to let us break ties between records that otherwise rank similarly. 

The [ranking criteria docs](https://www.algolia.com/doc/guides/managing-results/relevance-overview/in-depth/ranking-criteria/) explain all this in detail (and are worth reading), but the gist is that for a given query, a bunch of criteria -- text matching against attributes, synonyms, etc. -- are applied first, before custom ranking ever comes into play. Custom ranking is only applied when multiple records match a given query closely enough that Algolia can't tell which ones are "more relevant" than others. Otherwise, they don't carry much weight (if any at all).

For us, where this matters is for queries like "input" or "stack" that can match strongly for content in multiple categories. For unqualified queries like these, users will most likely be looking for Pulumi-specific content -- rather than, say, the API docs for the `aws.medialive.Input` or `aws.cloudformation.Stack` resources in the Registry (which would match for those terms). For this reason, we give docs a higher custom rank -- the assumption being that someone looking for docs about CloudFormation stacks will use additional words, like "cloudformation", to narrow the focus of those queries on their own. Also, because of how much _more_ content there is in the Registry, if we did rank Registry pages higher than docs, we'd be increasing the likelihood that our own docs content would be "drowned out" by the sheer volume of stuff that exists in the Registry; the more packages and resources we have, the more words there are to compete with our core terminology. Where legitimate ambiguity does exist -- e.g., for queries like "aws" or "kubernetes" or "awsx", where we can't really guess at what the user is looking for -- ranking docs above Registry is an overall safer bet -- especially now that we have those fine cloud-specific landing pages to help them find what they're looking for (and perhaps query more narrowly next time).

So all up, my hunch is that ranking docs above Registry is the right way to go -- but again, still learning, so I reserve the right to change my mind. 😇